pkg/keystone/initcontainer.go

```diff
@@ -52,6 +53,14 @@ func initContainer(init APIDetails) []corev1.Container {
 	envVars["DatabaseUser"] = env.SetValue(init.DatabaseUser)
 	envVars["DatabaseName"] = env.SetValue(init.DatabaseName)
 
+	var clientConfig string
+	if init.DatabaseUseTLS {
```

Comment:
When a CA certificate is not specified in the MySQL client configuration, but the MySQL server is configured with TLS, two things can happen depending on client's configuration.

- The client will fail to connect as the server's TLS certificate cannot be validated - This scenario will occur if the client is configured to enforce certificate validation. Without the CA certificate to verify the server's certificate, the client will reject the connection.

- The client will accept the server's TLS certificate blindly - This will occur if the client is configured to allow connections without validating the server's certificate. The connection can be established, but it is insecure, because there is no way how to validate server's certificate without the CA, and therefore the connection can be compromised by man-in-the-middle.

So, both (a) and (b) can be true as MySQL client can be configured in either way (some client's services this don't provide, as far as I remember memcached do not accept server's TLS handshake, when there's no way how to validate its certificate).

Reply:
@Deydra71 thanks for the explanation. Do we have preference over configuring the client for option a) or b) here?

Reply:
The preference is to have CA certs available everywhere, so the certs can always be verified. So the option to accept TLS connection blindly is discouraged.

Reply:
Sure, but this code now allows not to have CA configured for KeystoneAPI CR but have TLS certs on the mysql server. So my question here is: do we want keyston-operator to enforce option a) ? If so the I think we need to change this PR.

Reply:
Enabling TLS in PyMySQL enforces certificate verification, I'm not sure that we can disable this verification. 
I think ultimately, once all Openstack clients can connect to the database via TLS, when the galera CR is configured to use TLS, we should create database users with a 'REQUIRE SSL' grant that would only allow connection over TLS.

One potential issue is if keystone is not configured for TLS, but the database is and enforces connection over TLS. If so, keystone will never be able to connect.

Maybe for these two cases we could add a specific condition DatabaseReady like the MemcachedReady one? And give a specific error when a TLS config mismatch is detected by the keystone operator?

Reply:
> I think ultimately, once all Openstack clients can connect to the database via TLS, when the galera CR is configured to use TLS, we should create database users with a 'REQUIRE SSL' grant that would only allow connection over TLS.

I think this is a good idea.

> One potential issue is if keystone is not configured for TLS, but the database is and enforces connection over TLS. 

I think this is something that only the openstack-operator can check effectively as that is the only place where both the Gal;era and the KeystoneAPI CR is visible. We can add a validation webhook check there that rejects an OpenStackControlPlane CR that configures the Galera part with TLS but not the KeystoneAPI (and other service CR) part.

> Maybe for these two cases we could add a specific condition DatabaseReady like the MemcachedReady one? And give a specific error when a TLS config mismatch is detected by the keystone operator?

Can  keystone-operator really detect it? If so then sure we can add a condition for it.